File: Data Structures & Algorithms/interleaving-string/submission-4.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int m = s1.size();
        int n = s2.size();

        vector<vector<int>> dp(m + 1, vector<int>(n + 1, -1));
        return dfs(0, 0, s1, s2, s3, dp);
        
    }
    bool dfs(int i, int j, string& s1, string& s2, string& s3, vector<vector<int>>& dp){
        if (i + j == s3.size()) {
            if (i == s1.size() and j == s2.size()) {
                return true;
            }
        }
        if (dp[i][j] != -1) return dp[i][j];

        int k = i + j;
        if (i < s1.size() && s3[k] == s1[i]) {
            if (dfs(i + 1, j, s1, s2, s3, dp)) return true;
        }
        if (j < s2.size() && s3[k] == s2[j]) {
            if (dfs(i, j + 1, s1, s2, s3, dp)) return true;
        }
        dp[i][j] = false;
        return false;
    }
};

File: Data Structures & Algorithms/surrounded-regions/submission-4.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        R, C = len(board), len(board[0])

        def dfs(r, c):
            if not (0 <= r < R and 0 <= c < C and board[r][c] == "O"):
                return

            board[r][c] = "#"

            for dr, dc in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                curr_r, curr_c = r + dr, c + dc
                dfs(curr_r, curr_c)

        for i in range(R):
            dfs(i, 0)
            dfs(i, C - 1)

        for j in range(C):
            dfs(0, j)
            dfs(R - 1, j)

        for i in range(R):
            for j in range(C):
                if board[i][j] == "#":
                    board[i][j] = "O"
                elif board[i][j] == "O":
                    board[i][j] = "X"

